2.train-VAE/betavae_multilayer.py

```diff
@@ -0,0 +1,267 @@
+    avg_train_loss = train_loss / len(train_loader.dataset)
+    return avg_train_loss
+
+def train_vae(model, train_loader, optimizer, epochs):
+    """
+    Train the VAE model.
+
+    Args:
+        model (VAE): VAE model to be trained.
+        train_loader (DataLoader): DataLoader for the training data.
+        optimizer (Optimizer): Optimizer for the model.
+        epochs (int, optional): Number of training epochs. Defaults to 5.
+    Returns:
+        Training history (loss)
+    """
+    train_loss_history = []
+    model.train()
+    for epoch in range(epochs):
+        avg_train_loss = train_model(model, train_loader, optimizer)
+        train_loss_history.append(avg_train_loss)
+        print(f"Epoch {epoch}, Loss: {avg_train_loss}")
+
+    return train_loss_history
+
+
+def evaluate_vae(model, val_loader):
+    """
+    Evaluate the VAE model.
+
+    Args:
+        model (VAE): VAE model to be evaluated.
+        test_loader (DataLoader): DataLoader for the test data.
+
+    Returns:
+        Average loss over the test dataset.
+    """
+    model.eval()
+    val_loss = 0
+    with torch.no_grad():
+        for batch in val_loader:
+            data = batch[0]
+            recon, mu, log_var = model(data)
+            val_loss += model.loss_function(recon, data, mu, log_var).item()
+    return val_loss / len(val_loader.dataset)
+
+
+def compile_vae(model, train_loader, val_loader, test_loader, optimizer, epochs):
+    model.train()
+    train_loss_history = []
+    val_loss_history = []
+    test_loss_history = []
+
+    for epoch in range(epochs):
+
+        avg_train_loss = train_model(model, train_loader, optimizer)
+        train_loss_history.append(avg_train_loss)
+
+        avg_val_loss = evaluate_vae(model, val_loader)
+        val_loss_history.append(avg_val_loss)
+
+        avg_test_loss = evaluate_vae(model, test_loader)
+        test_loss_history.append(avg_test_loss)
+
+        print(f"Epoch {epoch+1}, Train Loss: {avg_train_loss}, Val Loss: {avg_val_loss}, Test Loss: {avg_test_loss}")
+
+    return train_loss_history, val_loss_history, test_loss_history
+
+
+
+def extract_latent_dimensions(model, data_loader, metadata):
+    """
+    Extract latent dimensions from the VAE model and save them with Model IDs.
+
+    Args:
+        model (VAE): Trained VAE model.
+        data_loader (DataLoader): DataLoader for the data.
+        metadata (DataFrame): Metadata containing Model IDs.
+
+    Returns:
+        DataFrame with latent dimensions and Model IDs.
+    """
+    model.eval()
+    latent_space = []
+    with torch.no_grad():
+        for batch in data_loader:
+            data = batch[0]
+            mu, _ = model.encode(data)
+            latent_space.append(mu.cpu().numpy())
+
+    latent_space = np.concatenate(latent_space, axis=0)
+    latent_df = pd.DataFrame(latent_space)
+    latent_df.insert(0, 'ModelID', metadata['ModelID'])
+
+    latent_df_dir = pathlib.Path("./results/latent_df_multi.parquet")
```

Comment:
consider passing this path in as an arg. Hardcoded paths are less flexible. What if you want to move where you write the data to in the future?